modules/apps/oauth2-provider/oauth2-provider-service/src/main/java/com/liferay/oauth2/provider/internal/upgrade/v3_0_1/OAuth2AuthorizationUpgradeProcess.java

```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.oauth2.provider.internal.upgrade.v3_0_1;
+
+import com.liferay.oauth2.provider.configuration.OAuth2ProviderConfiguration;
+import com.liferay.petra.string.StringBundler;
+import com.liferay.portal.kernel.module.configuration.ConfigurationProvider;
+import com.liferay.portal.kernel.upgrade.UpgradeProcess;
+import com.liferay.portal.kernel.util.Time;
+
+import java.sql.Date;
+import java.sql.PreparedStatement;
+
+/**
+ * @author Jonathan McCann
+ */
+public class OAuth2AuthorizationUpgradeProcess extends UpgradeProcess {
+
+	public OAuth2AuthorizationUpgradeProcess(
+		ConfigurationProvider configurationProvider) {
+
+		_configurationProvider = configurationProvider;
+	}
+
+	@Override
+	protected void doUpgrade() throws Exception {
+		OAuth2ProviderConfiguration oAuth2ProviderConfiguration =
+			_configurationProvider.getSystemConfiguration(
+				OAuth2ProviderConfiguration.class);
+
+		int expiredAuthorizationsAfterlifeDuration = Math.max(
+			oAuth2ProviderConfiguration.
+				expiredAuthorizationsAfterlifeDuration(),
+			0);
+
+		long expiredAuthorizationsAfterlifeDurationMillis =
+			expiredAuthorizationsAfterlifeDuration * Time.SECOND;
+
+		Date purgeDate = new Date(System.currentTimeMillis());
+
+		purgeDate.setTime(
+			purgeDate.getTime() - expiredAuthorizationsAfterlifeDurationMillis);
+
+		String sql = StringBundler.concat(
+			"delete OAuth2Authorization, OA2Auths_OA2ScopeGrants from ",
+			"OAuth2Authorization INNER JOIN OA2Auths_OA2ScopeGrants where ",
+			"(OAuth2Authorization.accessTokenExpirationDate < ?) and ",
+			"(((OAuth2Authorization.refreshTokenExpirationDate is not null) ",
+			"and (OAuth2Authorization.refreshTokenExpirationDate < ?)) or ",
+			"(OAuth2Authorization.refreshTokenExpirationDate is null)) and ",
+			"OAuth2Authorization.oAuth2AuthorizationId = ",
+			"OA2Auths_OA2ScopeGrants.oAuth2AuthorizationId");
```

Comment:
Hi @jonathanmccann, 

Do you think this single `DELETE` statement would be able to handle possibly millions of affected rows? We are a bit concerned that it might cause issues with "redo" logs of some DBs. Some batching might we required here or I thought about copying the still active rows to a temp table truncating the original table and copying them back but I'm not sure if it's possible here or can be implemented in a DMBS agnostic way. Can you consider this and write your opinion?

Thank you!

Reply:
@oliverkecskemety Thank you for taking a look at this.

In regards to batching, wouldn't that run into the same issue of filling up the logs? I'm not familiar with that type of database so I'm not sure what is being logged or how a single query would be handled differently from multiple smaller queries.

For truncating the table, it would depend on how many valid authorizations are present in the database. With this approach we'd need to insert every valid entry twice (once to the temporary table and then once back into the actual table). This would also need to be done for two tables so that will be four _inserts_ for every valid authorization.
Also, how do these logging databases handle a truncate? Would that be recorded as a single statement or would all of the millions of expired authorizations still be logged and the same issue would occur?

Reply:
Hi @jonathanmccann,

Totally valid questions but to be honest I'm not sure about the answers. I'll forward the PR to Team Uniform for some help regarding the SQL part as they should have already experienced such cases.